src/plugins/intel_cpu/src/nodes/vnode.cpp

```diff
@@ -0,0 +1,148 @@
+// Copyright (C) 2018-2023 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "vnode.h"
+
+#include "common/bfloat16.hpp"
+#include "common/cpu_memcpy.h"
+#include "cpu/x64/cpu_isa_traits.hpp"
+#include "shape_inference/shape_inference_internal_dyn.hpp"
+#include "utils/plain_tensor.hpp"
+
+#include <chrono>
+#include <string>
+#include <vector>
+
+namespace ov {
+namespace intel_cpu {
+namespace node {
+
+template <typename T>
+struct VNode::VNodeExecutorCausalMaskPreprocess : public VNode::Executor {
+    void execute(dnnl::stream strm,
+                 intel_cpu::Node * pnode,
+                 const intel_cpu::VNode::Config& config) override {
+        ov::intel_cpu::PlainTensor t_attention_mask(pnode->getSrcMemoryAtPort(0));
+        ov::intel_cpu::PlainTensor t_batch_size(pnode->getSrcMemoryAtPort(1));
+        ov::intel_cpu::PlainTensor t_cache_positions(pnode->getSrcMemoryAtPort(2));
+        ov::intel_cpu::PlainTensor t_kvLen(pnode->getSrcMemoryAtPort(3));
+
+        auto mask_length = t_attention_mask.size(-1);
+        auto batch_size = static_cast<size_t>(*t_batch_size.ptr<int32_t>(0));
+        auto kvLen = static_cast<size_t>(*t_kvLen.ptr<int32_t>(0));
+        auto qLen = t_cache_positions.size(0);
+
+        VectorDims newDims{batch_size, 1, qLen, kvLen};
+        pnode->redefineOutputMemory({newDims});
+        ov::intel_cpu::PlainTensor t_dst(pnode->getDstMemoryAtPort(0));
+
+        DEBUG_LOG("VNode::execute", config.type, "  batch_size=", batch_size, " qLen=", qLen, " kvLen=", kvLen);
+        DEBUG_LOG("VNode::execute  attention_mask=", t_attention_mask);
+        DEBUG_LOG("VNode::execute  cache_positions=", t_cache_positions);
+
+        /*
+            ##### preprocess
+            min_dtype = torch.finfo(dtype).min
+            causal_mask = self.causal_mask[None, None, :, :].repeat(batch_size, 1, 1, 1).to(dtype) * min_dtype
+            causal_mask = causal_mask.to(dtype=dtype, device=device)
+
+            mask_length = attention_mask.shape[-1]
+            padding_mask = causal_mask[..., :mask_length].eq(0.0) * attention_mask[:, None, None, :].eq(0.0)
+            causal_mask[..., :mask_length] = causal_mask[..., :mask_length].masked_fill(padding_mask, min_dtype)
+
+            ##### when being used will be further sliced
+            causal_mask = attention_mask
+                if attention_mask is not None and cache_position is not None:
+                    causal_mask = causal_mask[:, :, cache_position, : key_states.shape[-2]]
+        */
+
+        // raw_causal_mask is already ensured to be triu by transformation
+        // attention_mask  : i64[N, kv_len], 0 means mask-out, 1 means attends to
+        // dst [batch_size, 1, q_len, kvLen]
+        auto* prow = t_cache_positions.ptr<int32_t>(0);
+        T min_dtype = std::numeric_limits<T>::lowest();
+        for (size_t n = 0; n < batch_size; n++) {
```

Comment:
The casual mask generation could run in parallel in B and q_len, is it worthy to parallel the code?

Reply:
good idea, will fix!